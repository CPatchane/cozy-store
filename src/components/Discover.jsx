import React, { Component } from 'react'

import { translate } from 'cozy-ui/react/I18n'

import SmallAppItem from './SmallAppItem'

const DEMO_APPS = [
  {
    slug: 'bank',
    name: 'Bank',
    developer: {name: 'Cozy'},
    editor: 'Cozy',
    icon: require('../assets/temp_apps/icon-bank.svg')
  },
  {
    slug: 'sante',
    name: 'Sant√©',
    developer: {name: 'MAIF Ltd'},
    editor: '',
    icon: require('../assets/temp_apps/icon-sante.svg')
  },
  {
    slug: 'mon_logis',
    name: 'Mon Logis',
    developer: {name: 'HoodBrains'},
    editor: '',
    icon: require('../assets/temp_apps/icon-mon_logis.svg')
  },
  {
    slug: 'calendar',
    name: 'Calendar',
    developer: {name: 'Cozy'},
    editor: 'Cozy',
    icon: require('../assets/temp_apps/icon-calendar.svg')
  }
]

class Discover extends Component {
  render () {
    const { t } = this.props
    return (
      <div className='sto-discover sto-soon'>
        <h2 className='sto-discover-title'>{t('discover.title')}</h2>
        <div className='sto-discover-get-started'>
          <h3 className='sto-discover-get-started-title'>
            {t('discover.get_started')}
          </h3>
          <div className='sto-discover-get-started-list'>
            {DEMO_APPS.map(a => {
              return <SmallAppItem
                slug={a.slug}
                developer={a.developer}
                editor={a.editor}
                icon={a.icon}
                name={a.name}
                version={a.version}
              />
            })}
          </div>
        </div>
      </div>
    )
  }
}

export default translate()(Discover)
