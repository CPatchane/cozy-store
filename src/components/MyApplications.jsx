import React, { Component } from 'react'

import { Route } from 'react-router-dom'
import { translate } from 'cozy-ui/react/I18n'
import Spinner from 'cozy-ui/react/Spinner'

import SmallAppItem from './SmallAppItem'
import ApplicationModal from './ApplicationModal'

class MyApplications extends Component {
  constructor (props) {
    super(props)
    props.fetchApps()
  }

  onAppClick (appSlug) {
    this.props.history.push(`/myapps/${appSlug}/manage`)
  }

  render () {
    const { t, myApps, isFetching, error } = this.props
    return (
      <div className='sto-myapps'>
        <h2 className='sto-myapps-title'>{t('myapps.title')}</h2>
        <div className='sto-myapps-list'>
          {!isFetching && myApps && !!myApps.length &&
            myApps.map(a => {
              return <SmallAppItem
                slug={a.slug}
                developer={a.developer}
                editor={a.editor}
                icon={a.icon}
                name={a.name}
                version={a.version}
                onClick={() => this.onAppClick(a.slug)}
              />
            })
          }
          {error &&
            <p>{error}</p>
          }
        </div>
        {isFetching &&
          <Spinner
            size='xxlarge'
            loadingType='appsFetching'
            middle='true'
          />
        }

        <Route path='/myapps/:appSlug/manage' render={({ match }) =>
          (<ApplicationModal {...this.props} match={match} />)
        } />
      </div>
    )
  }
}

export default translate()(MyApplications)
