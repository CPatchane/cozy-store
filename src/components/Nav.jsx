import React from 'react'
import { translate } from 'cozy-ui/react/I18n'
import { NavLink } from 'react-router-dom'

export const Nav = ({ t }) => (
  <nav>
    <ul className='coz-nav'>
      <li className='coz-nav-item'>
        <NavLink to='/discover' className='sto-cat-discover coz-nav-link' activeClassName='active'>
          { t('nav.discover') }
        </NavLink>
      </li>
      <li className='coz-nav-item'>
        <NavLink to='/myapps' className='sto-cat-myapps coz-nav-link' activeClassName='active'>
          { t('nav.myapps') }
        </NavLink>
      </li>
    </ul>
  </nav>
)

export default translate()(Nav)
