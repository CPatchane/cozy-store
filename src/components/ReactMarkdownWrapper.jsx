import React from 'react'
import ReactMarkdown from 'react-markdown'

export const reactMarkdownRendererOptions = {
  Link: props => <a href={props.href} target='_blank'>{props.children}</a>
}

export const ReactMarkdownWrapper = ({ source }) =>
  <ReactMarkdown
    source={source}
    renderers={reactMarkdownRendererOptions}
  />

export default ReactMarkdownWrapper
