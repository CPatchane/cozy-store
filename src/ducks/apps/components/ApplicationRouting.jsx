import React, { Component } from 'react'
import { Route, withRouter } from 'react-router-dom'

import ApplicationDetails from './ApplicationDetails'
import UninstallModal from './UninstallModal'

export class ApplicationRouting extends Component {
  render () {
    const { installedApps, isFetching, parent, history } = this.props
    return (
      <div>
        <Route path={`/${parent}/:appSlug`} render={({ match }) => {
          if (isFetching) return
          if (installedApps.length && match.params) {
            const app = installedApps.find(app => app.slug === match.params.appSlug)
            if (!app) return history.push(`/${parent}`)
            return <ApplicationDetails app={app} parent={parent} />
          }
        }} />
        <Route path={`/${parent}/:appSlug/manage`} render={({ match }) => {
          if (isFetching) return
          if (installedApps.length && match.params) {
            const app = installedApps.find(app => app.slug === match.params.appSlug)
            if (!app) return history.push(`/${parent}`)
            return <UninstallModal uninstallApp={this.props.uninstallApp} parent={parent} error={this.props.actionError} app={app} match={match} />
          }
        }} />
      </div>
    )
  }
}

export default withRouter(ApplicationRouting)
