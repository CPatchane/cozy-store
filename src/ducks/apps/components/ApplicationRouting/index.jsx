import React, { Component } from 'react'
import { Route, withRouter } from 'react-router-dom'

import { translate } from 'cozy-ui/react/I18n'
import Alerter from 'cozy-ui/react/Alerter'

import UninstallModal from '../UninstallModal'
import InstallModal from '../InstallModal'

import AppRoute from './AppRoute'
import ChannelRoute from './ChannelRoute'
import PermissionsRoute from './PermissionsRoute'
import ConfigureRoute from './ConfigureRoute'

import { APP_TYPE } from 'ducks/apps'

export class ApplicationRouting extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isInstallSuccess: false
    }
  }

  componentWillReceiveProps(nextProps) {
    // don't react on error
    if (nextProps.actionError) return
    // on install success
    if (this.props.isInstalling && !nextProps.isInstalling) {
      const { location, parent, t } = this.props
      const pathRegex = new RegExp(`^/${parent}/([^/]*)/.*`)
      const matches = location.pathname.match(pathRegex)
      if (!matches || matches.length < 1) return this.redirectTo(`/${parent}/`)
      const app = this.getAppFromMatchOrSlug(null, matches[1])
      if (app.type === APP_TYPE.KONNECTOR) {
        return this.redirectTo(`/${parent}/${app.slug}/configure`)
      } else {
        Alerter.success(t('app_modal.install.message.install_success'), {
          duration: 3000
        })
        return this.redirectTo(`/${parent}/${app.slug}`)
      }
    } else if (this.props.isUninstalling && !nextProps.isUninstalling) {
      const { location, parent, t } = this.props
      const pathRegex = new RegExp(`^/${parent}/([^/]*)/.*`)
      const matches = location.pathname.match(pathRegex)
      if (!matches || matches.length < 1) return this.redirectTo(`/${parent}/`)
      const app = this.getAppFromMatchOrSlug(null, matches[1])
      Alerter.success(t('app_modal.uninstall.message.success'), {
        duration: 3000
      })
      return this.redirectTo(`/${parent}/${app.slug}`)
    }
  }

  getAppFromMatchOrSlug = (match, slug) => {
    const appsArray = this.props.apps || this.props.installedApps || []
    const appSlug = slug || (match && match.params && match.params.appSlug)
    if (!appsArray.length || !appSlug) return null
    const app = appsArray.find(app => app.slug === appSlug)
    return app
  }

  redirectTo = target => {
    return history.replace(target)
  }

  render() {
    const {
      fetchLatestApp,
      installApp,
      uninstallApp,
      updateApp,
      isFetching,
      isAppFetching,
      isInstalling,
      parent,
      history,
      actionError,
      fetchError
    } = this.props
    return (
      <div>
        <AppRoute
          getApp={this.getAppFromMatchOrSlug}
          isFetching={isFetching}
          parent={parent}
          redirectTo={this.redirectTo}
        />
        <ChannelRoute
          actionError={actionError}
          fetchError={fetchError}
          fetchLatestApp={fetchLatestApp}
          getApp={this.getAppFromMatchOrSlug}
          installApp={installApp}
          isAppFetching={isAppFetching}
          isFetching={isFetching}
          isInstalling={isInstalling}
          parent={parent}
          redirectTo={this.redirectTo}
          updateApp={updateApp}
        />
        <Route
          path={`/${parent}/:appSlug/manage`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            if (app && app.installed && !app.availableVersion) {
              if (!app.uninstallable)
                return history.replace(`/${parent}/${app.slug}`)
              return (
                <UninstallModal
                  uninstallApp={uninstallApp}
                  parent={`/${parent}`}
                  uninstallError={actionError}
                  app={app}
                />
              )
            } else {
              return (
                <InstallModal
                  installApp={installApp}
                  parent={`/${parent}`}
                  installError={actionError}
                  app={app}
                  isInstalling={isInstalling}
                />
              )
            }
          }}
        />
        <PermissionsRoute
          getApp={this.getAppFromMatchOrSlug}
          isFetching={isFetching}
          parent={parent}
          redirectTo={this.redirectTo}
        />
        <ConfigureRoute
          getApp={this.getAppFromMatchOrSlug}
          isFetching={isFetching}
          parent={parent}
          redirectTo={this.redirectTo}
        />
      </div>
    )
  }
}

export default withRouter(translate()(ApplicationRouting))
