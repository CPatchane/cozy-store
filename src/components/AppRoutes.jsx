import React from 'react'
import { Route, Switch, Redirect } from 'react-router-dom'

import MyApplications from '../containers/MyApplications'

const AppRoutes = () => {
  return <Switch>
    <Route path='/myapps' component={MyApplications} />
    <Redirect from='/' to='/myapps' />
  </Switch>
}

export default AppRoutes
