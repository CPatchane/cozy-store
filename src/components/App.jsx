
import React from 'react'

import Sidebar from './Sidebar'
import AppRoutes from './AppRoutes'

const App = () => (
  <div className='sto-wrapper coz-sticky'>
    <Sidebar />
    <main className='sto-content'>
      <AppRoutes />
    </main>
  </div>
)
export default App
