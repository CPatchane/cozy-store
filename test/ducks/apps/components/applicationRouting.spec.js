'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'
import { Route } from 'react-router-dom'

import { ApplicationRouting } from 'ducks/apps/components/ApplicationRouting'

import mockApps from '../_mockApps'

// const mockRegistyApps = mockApps.filter(app => app.isInRegistry).filter(app =>
// (Array.isArray(app.versions.stable) && !!app.versions.stable.length))
const mockInstalledApps = mockApps.filter(a => a.installed)

const getMockProps = (parent, installedApps = mockInstalledApps, apps = mockApps, isFetching = false) => ({
  apps,
  installedApps,
  isFetching,
  parent,
  history: { push: jest.fn() },
  uninstallApp: jest.fn(),
  installApp: jest.fn()
})

describe('ApplicationRouting component with ApplicationDetails', () => {
  it('should handle correctly if app found', () => {
    const mockProps = getMockProps('myapps')
    const component = shallow(
      <ApplicationRouting {...mockProps} />
    )
    const routes = component.find(Route)
    expect(routes.length).toBe(2)
    const routeToDetails = routes.nodes[0]
    // collect in mockApps is installed and isInRegistry
    const routeProps = { match: { params: { appSlug: 'collect' } } }
    const resultComponent = routeToDetails.props.render(routeProps)
    expect(resultComponent).toBeDefined()
    expect(resultComponent).toMatchSnapshot()
  })

  it('should correctly go to parent if app not found', () => {
    const parent = 'myapps'
    const mockProps = getMockProps(parent)
    const component = shallow(
      <ApplicationRouting {...mockProps} />
    )
    const routes = component.find(Route)
    expect(routes.length).toBe(2)
    const routeToDetails = routes.nodes[0]
    // collect in mockApps is installed and isInRegistry
    const routeProps = { match: { params: { appSlug: 'mock' } } }
    const resultComponent = routeToDetails.props.render(routeProps)
    expect(mockProps.history.push.mock.calls.length).toBe(1)
    expect(mockProps.history.push.mock.calls[0][0]).toBe(`/${parent}`)
    expect(resultComponent).toBeUndefined()
  })

  it('should not return anything if apps list is empty', () => {
    const mockProps = getMockProps('myapps', [], [])
    const component = shallow(
      <ApplicationRouting {...mockProps} />
    )
    const routes = component.find(Route)
    expect(routes.length).toBe(2)
    const routeToDetails = routes.nodes[0]
    // collect in mockApps is installed and isInRegistry
    const routeProps = { match: { params: { appSlug: 'collect' } } }
    const resultComponent = routeToDetails.props.render(routeProps)
    expect(resultComponent).toBeUndefined()
  })

  it('should not return anything if isFetching', () => {
    const mockProps = getMockProps('myapps', [], [], true)
    const component = shallow(
      <ApplicationRouting {...mockProps} />
    )
    const routes = component.find(Route)
    expect(routes.length).toBe(2)
    const routeToDetails = routes.nodes[0]
    // collect in mockApps is installed and isInRegistry
    const routeProps = { match: { params: { appSlug: 'collect' } } }
    const resultComponent = routeToDetails.props.render(routeProps)
    expect(resultComponent).toBeUndefined()
  })
})

describe('ApplicationRouting component with Modal', () => {
  it('should handle correctly if installed app found', () => {
    const mockProps = getMockProps('myapps')
    const component = shallow(
      <ApplicationRouting {...mockProps} />
    )
    const routes = component.find(Route)
    expect(routes.length).toBe(2)
    const routeToDetails = routes.nodes[1]
    // collect in mockApps is installed and isInRegistry
    const routeProps = { match: { params: { appSlug: 'collect' } } }
    const resultComponent = routeToDetails.props.render(routeProps)
    expect(resultComponent).toBeDefined()
    expect(resultComponent).toMatchSnapshot()
  })

  it('should handle correctly if uninstalled app found', () => {
    const mockProps = getMockProps('myapps')
    const component = shallow(
      <ApplicationRouting {...mockProps} />
    )
    const routes = component.find(Route)
    expect(routes.length).toBe(2)
    const routeToDetails = routes.nodes[1]
    // photos in mockApps is not installed and isInRegistry
    const routeProps = { match: { params: { appSlug: 'photos' } } }
    const resultComponent = routeToDetails.props.render(routeProps)
    expect(resultComponent).toBeDefined()
    expect(resultComponent).toMatchSnapshot()
  })

  it('should correctly go to parent if app not found', () => {
    const parent = 'myapps'
    const mockProps = getMockProps(parent)
    const component = shallow(
      <ApplicationRouting {...mockProps} />
    )
    const routes = component.find(Route)
    expect(routes.length).toBe(2)
    const routeToDetails = routes.nodes[1]
    // collect in mockApps is installed and isInRegistry
    const routeProps = { match: { params: { appSlug: 'mock' } } }
    const resultComponent = routeToDetails.props.render(routeProps)
    expect(mockProps.history.push.mock.calls.length).toBe(1)
    expect(mockProps.history.push.mock.calls[0][0]).toBe(`/${parent}`)
    expect(resultComponent).toBeUndefined()
  })

  it('should not return anything if apps list is empty', () => {
    const mockProps = getMockProps('myapps', [], [])
    const component = shallow(
      <ApplicationRouting {...mockProps} />
    )
    const routes = component.find(Route)
    expect(routes.length).toBe(2)
    const routeToDetails = routes.nodes[1]
    // collect in mockApps is installed and isInRegistry
    const routeProps = { match: { params: { appSlug: 'collect' } } }
    const resultComponent = routeToDetails.props.render(routeProps)
    expect(resultComponent).toBeUndefined()
  })

  it('should not return anything if isFetching', () => {
    const mockProps = getMockProps('myapps', [], [], true)
    const component = shallow(
      <ApplicationRouting {...mockProps} />
    )
    const routes = component.find(Route)
    expect(routes.length).toBe(2)
    const routeToDetails = routes.nodes[1]
    // collect in mockApps is installed and isInRegistry
    const routeProps = { match: { params: { appSlug: 'collect' } } }
    const resultComponent = routeToDetails.props.render(routeProps)
    expect(resultComponent).toBeUndefined()
  })
})
