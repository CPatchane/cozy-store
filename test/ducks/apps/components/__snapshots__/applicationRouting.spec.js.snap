// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationRouting component with ApplicationDetails should handle correctly if app found 1`] = `
<_translate
  app={
    Object {
      "developer": Object {
        "name": "Cozy",
      },
      "editor": "cozy",
      "icon": "<svg></svg>",
      "installed": true,
      "isInRegistry": true,
      "name": "Collect",
      "related": "http://collect.cozy.mock/",
      "slug": "collect",
      "uninstallable": false,
      "version": "3.0.0",
      "versions": Object {
        "beta": Array [
          "3.0.0",
        ],
        "dev": Array [
          "3.0.0",
        ],
        "stable": Array [
          "3.0.0",
        ],
      },
    }
  }
  parent="myapps"
/>
`;

exports[`ApplicationRouting component with UninstallModal should handle correctly if app found 1`] = `
<_translate
  app={
    Object {
      "developer": Object {
        "name": "Cozy",
      },
      "editor": "cozy",
      "icon": "<svg></svg>",
      "installed": true,
      "isInRegistry": true,
      "name": "Collect",
      "related": "http://collect.cozy.mock/",
      "slug": "collect",
      "uninstallable": false,
      "version": "3.0.0",
      "versions": Object {
        "beta": Array [
          "3.0.0",
        ],
        "dev": Array [
          "3.0.0",
        ],
        "stable": Array [
          "3.0.0",
        ],
      },
    }
  }
  error={undefined}
  match={
    Object {
      "params": Object {
        "appSlug": "collect",
      },
    }
  }
  parent="myapps"
  uninstallApp={[Function]}
/>
`;
